### ISSUE TITLE
BF16 inaccuracy of TensorRT 10.14.1 at Slice nodes

### ISSUE THREAD

**Issue by dangtruong-shopee** [NVIDIA/TensorRT#4718]
## Environment

**TensorRT Version**: 10.14.1

**NVIDIA GPU**: A30

**NVIDIA Driver Version**: 580.126.09

**CUDA Version**: 13.0

**CUDNN Version**: 9.8.0 

Operating System:

Python Version (if applicable): 3.12.3

Tensorflow Version (if applicable):

PyTorch Version (if applicable):

Baremetal or Container (if so, version):

## Relevant Files

Google Drive link: https://drive.google.com/drive/u/0/folders/1GbuWrsyY5gzFnzgd6ZbcPg6PTHXZloXt

## Steps To Reproduce

Run the following command to run the graph in bf16 TRT. The output file is already included in the Google Drive link for your reference.
`polygraphy run graph_final_baked.onnx --trt --bf16 --trt-outputs mark all --load-inputs layerwise_inputs.json --save-outputs 10_14_all_outputs.json`

Check the following tensors in 10_14_all_outputs.json: llm_lt_click_reqcate1_items_input/slice_tile_num:0, llm_lt_click_pcate_items_input/slice_tile_num:0, you will find that their values are 0:

Taken from 10_14_all_outputs.json for your reference: 
```
llm_lt_click_reqcate1_items_input/slice_tile_num:0 [dtype=float32, shape=(1, 1)] | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0), max=0 at (0, 0), avg-magnitude=0, p90=0, p95=0, p99=0
llm_lt_click_pcate_items_input/slice_tile_num:0 [dtype=float32, shape=(1, 1)] | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0), max=0 at (0, 0), avg-magnitude=0, p90=0, p95=0, p99=0
```

In actuality, their values are not 0, I have extracted those nodes into a subgraph at slice_subgraph.onnx. Strangely enough, when running this subgraph in isolation, the issue does not occur. Run the following command to verify:
`polygraphy run slice_subgraph.onnx --trt --bf16 --load-inputs layerwise_inputs.json --save-outputs slice_trt_outputs.json 2>/dev/null && polygraphy inspect data slice_trt_outputs.json --show-values`

Outputs of running slice_subgraph.onnx:
```
llm_lt_click_pcate_items_input/slice_tile_num:0 [dtype=float32, shape=(1, 1)] | Stats: mean=23, std-dev=0, var=0, median=23, min=23 at (0, 0), max=23 at (0, 0), avg-magnitude=23, p90=23, p95=23, p99=23
        [[23.]]
    llm_lt_click_reqcate1_items_input/slice_tile_num:0 [dtype=float32, shape=(1, 1)] | Stats: mean=34, std-dev=0, var=0, median=34, min=34 at (0, 0), max=34 at (0, 0), avg-magnitude=34, p90=34, p95=34, p99=34
        [[34.]]
```

Please help me check this issue. It is affecting the accuracy of our bf16 engine and we would like to find a solution for it. Thank you in advance.

**Comment by bparks-nv** [comment]
Hi @dangtruong-shopee, thanks for the clear steps to reproduce. I can reproduce the same issue on my end. We're still digging into the issue to find the cause of this behavior. I tried cutting down the model to a smaller subgraph, and I could only get as far as `cotrain_model/dcn/concat:0` while preserving the behavior of `llm_lt_click_reqcate1_items_input/slice_tile_num:0` becoming 0, so it seems that concat node is somehow necessary to see the issue. We are continuing to look at this internally.

**Comment by bparks-nv** [comment]
Hi @dangtruong-shopee, when running with `--trt-outputs mark all`, TensorRT will not be able to fuse layers together as usual so this is not always a good proxy for debugging end-to-end accuracy issues. For this network, it seems like running with `--fp16` rather than `--bf16` might provide better accuracy, as fp16 has better precision at the cost of lower dynamic range. 

When running end-to-end in bf16, we see a generated kernel containing the following onnx nodes results in all-zero outputs:
<details><summary>Click to expand list of onnx nodes</summary>

- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/Mean
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/Mean_1
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/Square
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/sub
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/Rsqrt
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/add
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l2_h_0/fully_connected/BiasAdd
- predicts/Sigmoid_2
- predicts/add_2
- predicts/sub_2
- predicts/ads_org_direct_order1d_predictions
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/mul
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/Relu
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l2_h_0/fully_connected/MatMul
- cotrain_model/MaskBlockMMoE/DNN_MaskBlockMMoE_taskid2_l1_h_0/fully_connected/Rsqrt__3847

</details>

We're also investigating what's happening with `mark all` outputs, but presumably the most important issue for you is end-to-end accuracy. 

Can you get an accurate end-to-end result when running with `--fp16` without using `--trt-outputs mark all`?

**Comment by dangtruong-shopee** [comment]
Hi @bparks-nv, yes we can achieve end-to-end accuracy when running with `--fp16`. We would like to have an option to use bf16 as well if possible. The use of `--trt-outputs mark all` was only our attempt to find out why bf16 wasn't behaving as expected because we do not know how else to investigate this. Are you able to find out what's wrong with the use of bf16?

**Comment by bparks-nv** [comment]
Hi @dangtruong-shopee we're still investigating what specifically is going wrong with the bf16. One possibility is a compilation bug, and another possibility is [catastrophic cancellation](https://en.wikipedia.org/wiki/Catastrophic_cancellation) dependent on the lower number of precision bits in bf16 (e.g. some intermediate tensors that are large and similar values get subtracted from each other; in bf16 they might round to the same value and subtract to 0).

I would suggest using fp16 in the mean time as a workaround, since it should have the same performance as bf16 and it sounds like it avoids the accuracy issue.

**Comment by dangtruong-shopee** [comment]
@bparks-nv noted on that. Please let me know if there are any updates regarding this issue. I'll relay your suggestion to my team. Thank you.
